[UnifiedPDF] VoiceOver does not recognize the password form

aprotyas opened the pull request:
#### 0d359e074687d590e5888bf9beb8d0fe5d4986b4
<pre>
[UnifiedPDF] VoiceOver does not recognize the password form
<a href="https://bugs.webkit.org/show_bug.cgi?id=297895">https://bugs.webkit.org/show_bug.cgi?id=297895</a>
<a href="https://rdar.apple.com/155907450">rdar://155907450</a>

Reviewed by Tyler Wilcock.

When WebPage reports the presence of a main frame plugin, we
unconditionally reach for the AX object wrapper corresponding to the
plugin. This has the undesired effect of stealing AX away from general
web content, which is usually fine except when said web content is
presented over a plugin, as is the case for the password form for locked
documents.

In this patch, we teach WKAccessibilityWebPageObject that it may be
necessary to disregard the plugin AX object even when a main frame
plugin is present. We go with this approach over teaching the plugin&apos;s
AX object (WKAccessibilityPDFDocumentObject) about the password
annotations because (a) the concept of an active annotation is still not
platform agnostic and (b) it would introduce an unfortunate coordinate
transformation spaghetti inside the accessibility object since it will
have to undo some transformations performed on the hit test point.

More details in-line.

* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::presentingLockedContent const):

Helper query that forwards PDFPluginBase::isLocked().

* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::accessibilityObjectForMainFramePlugin):

This method has the same definition on all Cocoa platforms, so we
deduplicate.

(WebKit::WebPage::fallbackToWebContentAXObjectForMainFramePlugin const):

Introduce a query that allows our AX wrapper object finding logic to
look at web content, rather than the plugin&apos;s bespoke object. You could
imagine this being necessary for multiple conditions, but the only one
it abstracts away for now is the presence of a password form (i.e. if
the document is locked)

* Source/WebKit/WebProcess/WebPage/WKAccessibilityWebPageObjectIOS.mm:
(-[WKAccessibilityWebPageObject accessibilityHitTest:]):

Multiple changes:
- Lift a clarifying comment from the macOS method definition about
  coordinate transformations for the plugin.
- Respect fallbackToWebContentAXObjectForMainFramePlugin and perform
  coordinate conversions regardless of main frame plugin presence as
  needed.
- Refactor the code so there is only one terminal call to hitTest:.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::accessibilityObjectForMainFramePlugin): Deleted.

Deduplicated this.

* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.h:
* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
(-[WKAccessibilityWebPageObjectBase fallbackToWebContentAXObjectForMainFramePlugin]):

Just forward WebPage::fallbackToWebContentAXObjectForMainFramePlugin().
Makes callsites a bit easier to read.

(-[WKAccessibilityWebPageObjectBase accessibilityRootObjectWrapper:]):

Respect fallbackToWebContentAXObjectForMainFramePlugin and return
general web content AX cache object when appropriate.

* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectMac.mm:
(-[WKAccessibilityWebPageObject accessibilityHitTest:]):

Same added logic and refactor as the iOS method definition.

* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::accessibilityObjectForMainFramePlugin): Deleted.

Deduplicated this.

Canonical link: <a href="https://commits.webkit.org/299180@main">https://commits.webkit.org/299180@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/66dceefd7075ffe0ad5abfa3ce0603a136d530ae

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118165 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37843 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28479 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124320 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70205 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b0874203-ef2f-4905-84dd-1dfa67816195) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38536 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46425 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89670 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59297 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/784110cf-0eec-4460-b431-9069bb4cee66) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121118 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30692 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105944 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70163 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e4fbb078-a1fd-4a5b-bd3d-f8289961e5f6) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29760 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67988 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100118 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24237 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127397 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45068 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33970 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98348 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45429 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102165 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98137 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24952 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43529 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/21517 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/41533 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44940 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50614 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44400 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47745 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46089 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebKit/WebProcess/Plugins/PluginView.cpp:
```diff
@@ -1241,6 +1241,14 @@ bool PluginView::pluginDelegatesScrollingToMainFrame() const
+    if (!m_isInitialized)
+        return false;
+
+    return m_plugin->isLocked();
```
This could also be written as:

```cpp
return m_isInitialized && m_plugin->isLocked();
```

Unless you feel the current version is more readable.

aprotyas replied:
This style follows the file convention, but I don't think your suggestion is any less readable, so I'll go with it

twilco commented on Source/WebKit/WebProcess/Plugins/PluginView.h:
```diff
@@ -169,6 +169,8 @@ class PluginView final : public WebCore::PluginViewBase {
 
     bool pluginDelegatesScrollingToMainFrame() const;
 
+    bool presentingLockedContent() const;
```
[WebKit coding guidelines](https://webkit.org/code-style-guidelines/#names-bool) recommend bool values (and I thought function names too, but the guidelines don't mention that) start with "is" / "has" / "did" / etc. So maybe `isPresentingLockedContent()`?

aprotyas replied:
Good point, I will prefix with `is`

twilco commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -1463,6 +1463,29 @@ static void drawPDFPage(PDFDocument *pdfDocument, CFIndex pageIndex, CGContextRe
+#if ENABLE(PDF_PLUGIN)
+    if (RefPtr pluginView = mainFramePlugIn())
+        return pluginView->accessibilityObject();
+#endif
```
Should the `if (!m_page)` check be moved into the `if ENABLE(PDF_PLUGIN)` block, since `!if ENABLE(PDF_PLUGIN)` always returns nil anyways?

aprotyas replied:
Sure

twilco commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -1245,6 +1245,7 @@ class WebPage final : public API::ObjectImpl<API::Object::Type::BundlePage>, pub
     void setIsolatedTree(Ref<WebCore::AXIsolatedTree>&&);
 #endif
     NSObject *accessibilityObjectForMainFramePlugin();
+    bool fallbackToWebContentAXObjectForMainFramePlugin() const;
```
I recognize this method name is already long, so will leave this as a judgement call for you to make. But in line with [code style guidelines](https://webkit.org/code-style-guidelines/#names-bool), maybe this should have a `should` prefix on the method name?

aprotyas replied:
Hmm... I started with `should` but dropped it for brevity, but maybe there's more virtue in following the guide.

twilco replied:
Yeah this is a tough one, either option is fine I think

twilco reviewed (approved):
r+ assuming you've manually tested main-frame PDF plugins and embedded `<embed>` style PDFs with VoiceOver

aprotyas committed:
[UnifiedPDF] VoiceOver does not recognize the password form
https://bugs.webkit.org/show_bug.cgi?id=297895
rdar://155907450

Reviewed by Tyler Wilcock.

When WebPage reports the presence of a main frame plugin, we
unconditionally reach for the AX object wrapper corresponding to the
plugin. This has the undesired effect of stealing AX away from general
web content, which is usually fine except when said web content is
presented over a plugin, as is the case for the password form for locked
documents.

In this patch, we teach WKAccessibilityWebPageObject that it may be
necessary to disregard the plugin AX object even when a main frame
plugin is present. We go with this approach over teaching the plugin's
AX object (WKAccessibilityPDFDocumentObject) about the password
annotations because (a) the concept of an active annotation is still not
platform agnostic and (b) it would introduce an unfortunate coordinate
transformation spaghetti inside the accessibility object since it will
have to undo some transformations performed on the hit test point.

More details in-line.

* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::presentingLockedContent const):

Helper query that forwards PDFPluginBase::isLocked().

* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::accessibilityObjectForMainFramePlugin):

This method has the same definition on all Cocoa platforms, so we
deduplicate.

(WebKit::WebPage::fallbackToWebContentAXObjectForMainFramePlugin const):

Introduce a query that allows our AX wrapper object finding logic to
look at web content, rather than the plugin's bespoke object. You could
imagine this being necessary for multiple conditions, but the only one
it abstracts away for now is the presence of a password form (i.e. if
the document is locked)

* Source/WebKit/WebProcess/WebPage/WKAccessibilityWebPageObjectIOS.mm:
(-[WKAccessibilityWebPageObject accessibilityHitTest:]):

Multiple changes:
- Lift a clarifying comment from the macOS method definition about
  coordinate transformations for the plugin.
- Respect fallbackToWebContentAXObjectForMainFramePlugin and perform
  coordinate conversions regardless of main frame plugin presence as
  needed.
- Refactor the code so there is only one terminal call to hitTest:.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::accessibilityObjectForMainFramePlugin): Deleted.

Deduplicated this.

* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.h:
* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectBase.mm:
(-[WKAccessibilityWebPageObjectBase fallbackToWebContentAXObjectForMainFramePlugin]):

Just forward WebPage::fallbackToWebContentAXObjectForMainFramePlugin().
Makes callsites a bit easier to read.

(-[WKAccessibilityWebPageObjectBase accessibilityRootObjectWrapper:]):

Respect fallbackToWebContentAXObjectForMainFramePlugin and return
general web content AX cache object when appropriate.

* Source/WebKit/WebProcess/WebPage/mac/WKAccessibilityWebPageObjectMac.mm:
(-[WKAccessibilityWebPageObject accessibilityHitTest:]):

Same added logic and refactor as the iOS method definition.

* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::accessibilityObjectForMainFramePlugin): Deleted.

Deduplicated this.

Canonical link: https://commits.webkit.org/299180@main
